### PR TITLE
RavenDB-14634 various fixes

### PR DESCRIPTION
**Pull request by karmeli87**

### CODE REVIEW

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 1]
```diff
@@ -2585,7 +2585,7 @@ private async Task<(long Index, object Result)> SendToLeaderAsyncInternal(Comman
         private async Task<(long Index, object Result)> SendToLeaderAsyncInternal(TransactionOperationContext context, CommandBase cmd)
         {
             //I think it is reasonable to expect timeout twice of error retry
-            var timeoutTask = TimeoutManager.WaitFor(Engine.OperationTimeout, _shutdownNotification.Token);
+            var timeoutTask = TimeoutManager.WaitFor(TimeSpan.FromMinutes(60), _shutdownNotification.Token);
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs]: Why is that?

Comment (karmeli87) [reply src/Raven.Server/ServerWide/ServerStore.cs]: we try to hunt why we got random timeout failures that cause to 4-5 tests simultaneously to fail

Comment (ppekrol) [reply src/Raven.Server/ServerWide/ServerStore.cs]: Something that @karmeli87 forgot to write here. This PR will not be merged. Just a pure bug hunting code.

@karmeli87 maybe track how much time we wait with stopwatch and if it is > Engine.OperationTimeout do a console writeline with some debug info? Thoughts?

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Client/Documents/Operations/DatabaseHealthCheckOperation.cs [hunk 2]
```diff
@@ -14,11 +15,16 @@ public RavenCommand GetCommand(DocumentConventions conventions, JsonOperationCon
 
         private class DatabaseHealthCheckCommand : RavenCommand
         {
+            public DatabaseHealthCheckCommand()
+            {
+                Timeout = TimeSpan.FromSeconds(15); // maybe even less?
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/DatabaseHealthCheckOperation.cs:20]: Shouldn't we have a single location to define the default timeout?

File: src/Raven.Server/Commercial/LicenseManager.cs [hunk 3]
```diff
@@ -468,7 +468,7 @@ public static LicenseStatus GetLicenseStatus(License license)
             return licenseStatus;
         }
 
-        public void TryActivateLicense(bool throwOnActivationFailure)
+        public async ValueTask TryActivateLicenseAsync(bool throwOnActivationFailure)
```
Comment (ayende) [inline src/Raven.Server/Commercial/LicenseManager.cs]: Why is this an `async ValueTask`?

File: src/Raven.Server/Rachis/RachisLogHistory.cs [hunk 4]
```diff
@@ -379,6 +379,18 @@ private static unsafe string ReadGuid(Table.TableValueHolder entryHolder)
             return guid;
         }
 
+        public bool ContainsGuid(ClusterOperationContext context, string guid)
```
Comment (ayende) [inline src/Raven.Server/Rachis/RachisLogHistory.cs]: Guid is a specific term, but you are passing just ids here, which can be any strings

File: src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs [hunk 5]
```diff
@@ -322,21 +322,45 @@ private void LogMessage(string message, Exception e = null, string database = nu
...
+                        }
+
+                        await _engine.PutAsync(cmd);
+                    }
                 }
             }
         }
 
+
+        private static string GetGuid(Dictionary<string, long> dic)
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs]: This is not a guid.

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 6]
```diff
@@ -2359,15 +2359,15 @@ public void AssignNodesToDatabase(ClusterTopology clusterTopology, DatabaseRecor
             return SendToLeaderAsync(addDatabaseCommand);
         }
 
-        public async Task EnsureNotPassiveAsync(string publicServerUrl = null, string nodeTag = "A", bool skipLicenseActivation = false)
+        public async ValueTask EnsureNotPassiveAsync(string publicServerUrl = null, string nodeTag = "A", bool skipLicenseActivation = false)
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs]: Why?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Operations/DatabaseHealthCheckOperation.cs (hunk 2 above)
Comment (ppekrol) [reply src/Raven.Client/Documents/Operations/DatabaseHealthCheckOperation.cs:20]: No, that would affect too many things. All internal'ish commands have this timeout set manually, In theory we could introduce sth like RavenInternalCommand which would set it by default.

File: src/Raven.Server/Rachis/RachisLogHistory.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Server/Rachis/RachisLogHistory.cs]: Okay, so would something like: `CommandUniqueId` would be a more proper term?

File: src/Raven.Server/Commercial/LicenseManager.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Server/Commercial/LicenseManager.cs]: In the common case `TryActivateLicenseAsync` just like `EnsureNotPassiveAsync` are expected to return without actually doing any `async` work.

File: src/Raven.Server/Rachis/RachisLogHistory.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Rachis/RachisLogHistory.cs]: Yes

File: src/Raven.Server/Commercial/LicenseManager.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/Commercial/LicenseManager.cs]: See here: https://blog.marcgravell.com/2019/08/prefer-valuetask-to-task-always-and.html

In general, if we have to call `await` internally, that isn't saving a lot.
And if we are going to await internally and outside, there is little point in here.
Also, this isn't likely to be in anything that is in the hot path

Comment (karmeli87) [reply src/Raven.Server/Commercial/LicenseManager.cs]: I read this and others few times. TBH I'm still confused :-)

As I understand, the bottom line of that post is to use `ValueTask` _always_ (it is even in the title). But to be aware of the API changes.

Yet, as long as `AsyncTaskMethodBuilder` and friends aren't showing up in the profiler `Task` should be fine to use.

Comment (ayende) [reply src/Raven.Server/Commercial/LicenseManager.cs]: In general, in our internal API, prefer `Task` unless we know that:

* This is called frequently.
* We expect (and will write code) to handle the completion in a synchronous manner.

For external API:

* If they are likely to complete with no async work, use `ValueTask`

**Pull request merged**